Implement "glue" for prover modules over koalabear field (old)

OmegaTymbJIep opened the pull request:
## Changes

- Created new `FlattenColumn` component in `common` package that:
   - Manages flattened limbs and mask columns
   - Implements projection constraints to verify data consistency
   - Supports reuse across multiple circuits to avoid duplication
- `ecarith`, `modexp`, `ecpair`
   - Used Flatten column for a Plonk in Wizard part.
   - Adjusted constraints by dividing 128-bit limbs into 8 columns of 16 bits.

### Checklist

* [ ] I wrote new tests for my new core changes.
* [ ] I have successfully ran tests, style checker and build against my new changes locally.
* [ ] I have informed the team of any breaking changes if there are any.

OmegaTymbJIep committed:
Implement "glue" for `ecarith` module over koalabear field (#3)

## Changes

Added comprehensive implementation of flatten column functionality to handle limb data in a circuit:

1. Created new `FlattenColumn` component in `common` package that:
   - Manages flattened limbs and mask columns
   - Implements projection constraints to verify data consistency
   - Supports reuse across multiple circuits to avoid duplication

2. Refactored EC arithmetic modules to use shared flatten column:
   - Updated ECADD and ECMUL to use common implementation
   - Added proper column sharing between circuits
   - Improved variable naming for clarity

olegfomenko committed:
Implement "glue" for modexp module over koalabear field (#2)

1. Flatten columns for limbs and circuit masks were used in the modexp module for both 256-bit and 4096-bit instances.
2. Updated modexp constraints for a limbs column divided into 8 smaller columns.

AlexandreBelling commented on prover/zkevm/prover/common/flatten_column.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// Assign maps trace limb columns and mask into the flattened columns.
+func (l *FlattenColumn) Assign(run *wizard.ProverRuntime) {
```
Minor note; if you rename the function [Run], it will satisfy the [wizard.ProverAction] interface. It can allow registering a flatten_column for rounds that are larger than 0. For the glue it won't matter but it is an easy way to extend the functionality

AlexandreBelling commented on prover/zkevm/prover/ecarith/ecadd.go:
```diff
@@ -63,22 +74,26 @@ func newEcAdd(comp *wizard.CompiledIOP, limits *Limits, src *EcDataAddSource, pl
 type EcDataAddSource struct {
 	CsEcAdd ifaces.Column
-	Limb    ifaces.Column
+	Limbs   [nbLimbsCols]ifaces.Column
```
It would make sense to a list of constants in `zkevm/prover/common` for this. Otherwise, we will need to redeclare them everywhere. Something like

```
const (
       NbLimbU32 int = 2
       NbLimbU64 int = 4
       NbLimbEthAddress = 10
       NbLimbU128 = 8
       NbLimbU256 = 16
)
```

It should be possible (I don't know if it is worth it) to implements special column types (also in `common`) with associated `VectorBuilder` types. There is already a `HiLo` type which is just a tuple of `U128`. If we push the idea further, it should be possible to automate many of the changes.

AlexandreBelling commented:
The PR is good, everything seems to make sense to me. Only had 2 suggestions.

OmegaTymbJIep committed:
Rename FlattenColumn method Assign to Run, add and use constants for limbs nb

Nazarevsky committed:
Remove redundant for loop

olegfomenko committed:
fixed modexp circuit limbSizeBits

OmegaTymbJIep committed:
Add flatten column tests with fixes

OmegaTymbJIep committed:
Use limbs slice in the generic bytes module

OmegaTymbJIep committed:
Update tests for generic bytes module

OmegaTymbJIep committed:
Implement "glue" for ecpair module over koalabear field (#5)

## Changes 

1. Use Flatten column in the ecpair module for a Plonk in Wizard part.
2. Adjust constraints for 8 columns of 16 bits.
3. Change the G2GroupCheckInstance.IsSuccess, so it uses 8 field values to represent the boolean, but the first 7 are empty.

Nazarevsky committed:
Fix modexp test

Nazarevsky committed:
Implement "glue" for ecdsa module over koalabear

* Make 16 bit limbs for address

* Make 16 bit limbs for txn signature

* Rewrite ecdata for 16 bit limbs

* Apply 16 bit divisions for unaligned data

* Fix address constrains, remove unused variables

* Break into 16 bit limbs

* Refactor constant name and structures assignment

* Refactor constant name and structures assignment

* Fix divideBytes empty input function, add test

* Fix wrong address trimming

* Change reader to read antichamber csv file

* Remove mocked data for modules that use hash module

* Add working ecdsa module

* Fix constraints and circuit

* Apply constraint for an every column

* Get rid of magic variables

* Fix address constraint over all columns

* Use constants from the common module

* Change constants, change comments, bring commented constraints back

* Add import for circuit

* Change constant name, remove commented import

* Fix ScanStreams, change limb size

* Make constant global

* Add left alignment for limbs

* Add GnarkDataLA column

* Uncomment projection query

* Remove commented print, fix formatting

* Rename DivideBytes into SplitBytes

* Fix magic values

* Change gnarkDataOffsetAlignment from bytes to bits

OmegaTymbJIep commented:
Moved to #8
